<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Part 1. Data preprocessing</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Part 1. Data preprocessing</h1>



<p>In this part we show data pre-processing steps. It includes quality assessment (QC), Trimming, sequences alignment and creating a count table.<br />
Most tools used here are packed in Docker containers. To see original commands for each tool check our <a href="https://github.com/anabeloff/docker_images">GitHub Docker Images repository</a>.</p>
<div id="qc-and-trimm" class="section level2">
<h2>QC AND TRIMM</h2>
<p>The Docker image ‘qualim’ contains FastQC and Trimmomatic.<br />
To use the following code first we define variables for the in-container script ‘quali.sh’.</p>
<p>Primary variable <code>ANALYSIS</code> has two options <code>trimm</code> and <code>fastqc</code>. First runs Trimmomatic then FastQC, while second just runs FastQC.<br />
Place all FASTQ files in the working directory. The script detects files with ‘.gz’ extension.</p>
<p>An example run:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="ex">docker</span> run --rm --name=<span class="st">&quot;qc&quot;</span> \</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">          -e ANALYSIS=<span class="st">&quot;fastqc&quot;</span> \</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">          --mount <span class="st">'type=bind,source=path/workingdir,target=/usr/local/src/rnaSeq/workingdrive'</span> \</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">          qualim</a></code></pre></div>
<p><strong>IMPORTANT</strong><br />
Do not perform ‘trimm’ on original FASTQ files. They will be deleted. When using ‘trimm’ option with container based on ‘qualim’ <strong>REMEMBER</strong>, by default after trimming it will <strong>DELETE</strong> all original fastq and unpaired files. It leaves only paired reads files.<br />
This feature was created to optimise the use of disc space on cloud platforms.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="ex">docker</span> run --rm --name=<span class="st">&quot;trim&quot;</span> \</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">          -e THREADS=6 \</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">          -e CROP_LEN=110 \</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">          -e MIN_LEN=99 \</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">          -e HEADCROP=10 \</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">          -e ANALYSIS=<span class="st">&quot;trimm&quot;</span> \</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">          --mount <span class="st">'type=bind,source=path/workingdir,target=/usr/local/src/rnaSeq/workingdrive'</span> \</a>
<a class="sourceLine" id="cb2-8" data-line-number="8">          qualim</a></code></pre></div>
</div>
<div id="star-alignment" class="section level2">
<h2>STAR ALIGNMENT</h2>
<p>To align with STAR (v.2.7) we first need to create a genome index. This step needs to be done only once.<br />
Because it is not a routine step we use interactive mode with STAR docker image.<br />
To run the following command, save in the working directory a genome FASTA file and corresponding GTF annotation.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="ex">docker</span> run -it --mount <span class="st">'type=bind,source=path/workingdir,target=/usr/local/src/rnaSeq/workingdrive'</span> staraligner /bin/bash</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co"># STAR index</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4">  <span class="ex">STAR</span> \</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">    --runThreadN 8 \</a>
<a class="sourceLine" id="cb3-6" data-line-number="6">    --runMode genomeGenerate \</a>
<a class="sourceLine" id="cb3-7" data-line-number="7">    --genomeSAindexNbases 11 \</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">    --genomeDir starIndex \</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">    --genomeFastaFiles cp_genome.fa \</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">    --sjdbGTFfile cp_genome.gtf</a></code></pre></div>
<p>On clound platforms (AWS) preferably use memory optimised instances.<br />
Alignment of large FASTQ files will require 16-32Gb of RAM (fungal genome).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="ex">docker</span> run --detach\</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">-e THREADS=8 \</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">--mount <span class="st">'type=bind,source=path/workingdir,target=/seqdata/workingdir'</span> \</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">staraligner /seqdata/star_InContainer_v2.sh *_paired.fastq.gz</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co"># Find container ID</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="ex">docker</span> ps</a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co"># Follow stdout in detached container</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="ex">docker</span> logs -f <span class="op">&lt;</span>container ID<span class="op">&gt;</span></a></code></pre></div>
</div>
<div id="se-object" class="section level2">
<h2>SE OBJECT</h2>
<p>On this step we create a count table as SummarizedExperiment object.<br />
Launch the container in the same directory with BAM files from the previous step.</p>
<div id="annotation" class="section level4">
<h4>Annotation</h4>
<p>First we create annotation data base from GFF file.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">library</span>(<span class="st">&quot;Rsamtools&quot;</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw">library</span>(<span class="st">&quot;GenomicFeatures&quot;</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw">library</span>(<span class="st">&quot;GenomicAlignments&quot;</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"></a>
<a class="sourceLine" id="cb5-5" data-line-number="5">GFFFILE &lt;-<span class="st"> &quot;cp_genome.gff&quot;</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6">SPECIES_NAME &lt;-<span class="st"> &quot;Cryphonectria parasitica&quot;</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7">TXDB_FILE &lt;-<span class="st"> &quot;crypa_annotation.sqlite&quot;</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"></a>
<a class="sourceLine" id="cb5-9" data-line-number="9">Format &lt;-<span class="st"> &quot;gff3&quot;</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"></a>
<a class="sourceLine" id="cb5-11" data-line-number="11">transdb&lt;-<span class="kw">makeTxDbFromGFF</span>(GFFFILE, <span class="dt">format =</span> Format, <span class="dt">organism =</span> SPECIES_NAME)</a>
<a class="sourceLine" id="cb5-12" data-line-number="12"></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="kw">saveDb</span>(transdb, TXDB_FILE) </a></code></pre></div>
<p>Create SummarizedExperiment Object.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">library</span>(<span class="st">&quot;Rsamtools&quot;</span>)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">library</span>(<span class="st">&quot;GenomicFeatures&quot;</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="kw">library</span>(<span class="st">&quot;GenomicAlignments&quot;</span>)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"></a>
<a class="sourceLine" id="cb6-5" data-line-number="5">SE_NAME &lt;-<span class="st"> &quot;se.RData&quot;</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6">TXDB_FILE &lt;-<span class="st"> &quot;crypa_annotation.sqlite&quot;</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="co"># Create BAM files list</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9">filenames &lt;-<span class="kw">list.files</span>(<span class="st">&quot;.&quot;</span>, <span class="dt">recursive=</span><span class="ot">TRUE</span>, <span class="dt">pattern=</span><span class="st">&quot;.bam$&quot;</span>, <span class="dt">full=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-10" data-line-number="10">bamfiles &lt;-<span class="st"> </span><span class="kw">BamFileList</span>(filenames, <span class="dt">yieldSize=</span><span class="dv">200000</span>)</a>
<a class="sourceLine" id="cb6-11" data-line-number="11"></a>
<a class="sourceLine" id="cb6-12" data-line-number="12"><span class="co">#Read annotation DB</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13">transdb &lt;-<span class="st"> </span><span class="kw">loadDb</span>(TXDB_FILE)</a>
<a class="sourceLine" id="cb6-14" data-line-number="14"></a>
<a class="sourceLine" id="cb6-15" data-line-number="15">genes &lt;-<span class="st"> </span>GenomicFeatures<span class="op">::</span><span class="kw">genes</span>(transdb) </a>
<a class="sourceLine" id="cb6-16" data-line-number="16"></a>
<a class="sourceLine" id="cb6-17" data-line-number="17">se &lt;-<span class="st"> </span><span class="kw">summarizeOverlaps</span>(<span class="dt">features=</span>genes, <span class="dt">reads=</span>bamfiles,</a>
<a class="sourceLine" id="cb6-18" data-line-number="18">                        <span class="dt">mode=</span><span class="st">&quot;Union&quot;</span>,</a>
<a class="sourceLine" id="cb6-19" data-line-number="19">                        <span class="dt">singleEnd=</span><span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb6-20" data-line-number="20">                        <span class="dt">ignore.strand=</span><span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb6-21" data-line-number="21">                        <span class="dt">fragments=</span><span class="ot">TRUE</span> )</a>
<a class="sourceLine" id="cb6-22" data-line-number="22"></a>
<a class="sourceLine" id="cb6-23" data-line-number="23"></a>
<a class="sourceLine" id="cb6-24" data-line-number="24"><span class="kw">saveRDS</span>(se, <span class="dt">file =</span> SE_NAME)</a></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
